// _js-base@0.0.1-beta@js-base/dist/DbCommand.js
var Util = require("util");
var Url = require("url");
var Path = require("path");
var ArgumentNullError_1 = require("./ArgumentNullError");
var DbCommand = (function () {
    function DbCommand(dataPath, dbName, entityName, entityId, query) {
        if (!dataPath)
            throw new ArgumentNullError_1.ArgumentNullError("dataPath");
        if (!dbName)
            throw new ArgumentNullError_1.ArgumentNullError("dbName");
        this.dataPath = dataPath;
        this.dbName = dbName;
        this.entityName = entityName;
        this.entityId = entityId;
        this.query = query;
    }
    DbCommand.prototype.getDbRootPath = function () {
        return Path.join(this.dataPath, this.dbName);
    };
    DbCommand.prototype.getEntityRootPath = function () {
        return Path.join(this.getDbRootPath(), this.entityName);
    };
    DbCommand.prototype.getEntityPath = function () {
        return Path.join(this.getEntityRootPath(), this.getFilenameForEntity(this.entityId));
    };
    DbCommand.prototype.getFilenameForEntity = function (id) {
        return Util.format("%s.json", id);
    };
    DbCommand.prototype.hasEntityName = function () {
        return (this.entityName != null);
    };
    DbCommand.prototype.hasEntityId = function () {
        return (this.hasEntityName() && this.entityId != null);
    };
    DbCommand.prototype.forEntityId = function (id) {
        return new DbCommand(this.dataPath, this.dbName, this.entityName, id, this.query);
    };
    DbCommand.parseRequest = function (dataPath, request) {
        var url = Url.parse(request.url);
        var match = /^\/?([^/?]+)(?:\/([^/?]+))?(?:\/([^/?]+))?/.exec(url.path);
        if (match != null) {
            var dbName = match[1];
            var entityName = match[2];
            var entityId = match[3];
            var dbCommand = new DbCommand(dataPath, dbName, entityName, entityId, url.query);
            return dbCommand;
        }
    };
    return DbCommand;
})();
exports.DbCommand = DbCommand;
