// _js-base@0.0.1-beta@js-base/dist/DbProcessor.js
var FS = require("fs");
var Path = require("path");
var mkdirp = require("mkdirp");
var Uuid = require("node-uuid");
var PersistenceError_1 = require("./PersistenceError");
var DbProcessor = (function () {
    function DbProcessor(config) {
        this.config = config;
    }
    DbProcessor.prototype.getSingle = function (dbCommand, callback) {
        FS.exists(dbCommand.getEntityPath(), function (exists) {
            if (!exists)
                return callback(null);
            FS.readFile(dbCommand.getEntityPath(), function (err, entityData) {
                if (err)
                    return callback(null, err);
                callback(entityData.toString());
            });
        });
    };
    DbProcessor.prototype.getMany = function (dbCommand, search, callback) {
        var errors = new Array();
        FS.readdir(dbCommand.getEntityRootPath(), function (err, filenames) {
            if (err)
                return callback(null, errors, err);
            var results = new Array();
            var readFile = function (index) {
                if (index === void 0) { index = 0; }
                var filename = filenames[index];
                try {
                    var relativeFilePath = Path.join(dbCommand.getEntityRootPath(), filename);
                    FS.readFile(relativeFilePath, function (err, buffer) {
                        if (err)
                            throw Error(err);
                        var entityData = buffer.toString();
                        if (search == null || search.test(JSON.parse(entityData)))
                            results.push(entityData);
                        if (index < (filenames.length - 1))
                            readFile(++index);
                        else
                            callback(results, errors);
                    });
                }
                catch (err) {
                    var id = (new RegExp("(.+)\.[^\.]+$").exec(filename))[1];
                    errors.push(new PersistenceError_1.PersistenceError(id, err.toString()));
                }
            };
            readFile();
        });
    };
    DbProcessor.prototype.saveSingle = function (dbCommand, entity, callback) {
        dbCommand.entityId = (dbCommand.entityId || entity.id || Uuid.v1());
        entity.id = dbCommand.entityId;
        mkdirp(dbCommand.getEntityRootPath(), function (err) {
            if (err)
                return callback(null, err);
            var entityData = JSON.stringify(entity);
            FS.writeFile(dbCommand.getEntityPath(), entityData, function (err) {
                if (err)
                    throw Error(err);
                callback(entityData);
            });
        });
    };
    DbProcessor.prototype.saveMany = function (dbCommand, entities, callback) {
        var _this = this;
        var count = entities.length;
        var createdEntities = new Array();
        var errors = new Array();
        if (count == 0)
            return callback(createdEntities, errors);
        var save = function (index) {
            if (index === void 0) { index = 0; }
            var entity = entities[index];
            try {
                _this.saveSingle(dbCommand.forEntityId(entity.id), entity, function (createdEntityData) {
                    createdEntities.push(createdEntityData);
                    if (index < (count - 1))
                        save(index + 1);
                    else
                        callback(createdEntities, errors);
                });
            }
            catch (err) {
                errors.push(new PersistenceError_1.PersistenceError(entity.id, err));
            }
        };
        save();
    };
    DbProcessor.prototype.deleteSingle = function (dbCommand, callback) {
        var entityPath = dbCommand.getEntityPath();
        FS.exists(entityPath, function (exists) {
            if (!exists)
                return callback();
            FS.unlink(entityPath, function (err) {
                if (err)
                    return callback(err);
                callback();
            });
        });
    };
    DbProcessor.prototype.deleteMany = function (dbCommand, ids, callback) {
        var _this = this;
        var count = ids.length;
        var deletedIds = new Array();
        var errors = new Array();
        if (count == 0)
            return callback(deletedIds, errors);
        var save = function (index) {
            if (index === void 0) { index = 0; }
            var id = ids[index];
            try {
                _this.deleteSingle(dbCommand.forEntityId(id), function () {
                    deletedIds.push(id);
                    if (index < (count - 1))
                        save(index + 1);
                    else
                        callback(deletedIds, errors);
                });
            }
            catch (err) {
                errors.push(new PersistenceError_1.PersistenceError(id, err));
            }
        };
        save();
    };
    return DbProcessor;
})();
exports.DbProcessor = DbProcessor;
