// _js-base@0.0.1-beta@js-base/dist/SearchCriteria.js
var SearchOperatorRegistry_1 = require("./SearchOperatorRegistry");
var SearchCriteria = (function () {
    function SearchCriteria(fieldName, operator, rawValue, invert) {
        if (invert === void 0) { invert = false; }
        this.invert = false;
        var resolvedOperator = SearchOperatorRegistry_1.SearchOperatorRegistry.resolve(operator);
        var value = resolvedOperator.parseValue(rawValue);
        this.fieldName = fieldName;
        this.operator = resolvedOperator;
        this.value = value;
        this.invert = invert;
    }
    SearchCriteria.prototype.validate = function () {
        if (this.fieldName == null)
            throw Error("Search criteria has null field name.");
        if (this.operator == null)
            throw Error("Search criteria has null operator.");
        this.operator.validateValue(this.value);
    };
    ;
    SearchCriteria.prototype.test = function (entity) {
        var result = this.operator.exec(entity, this.fieldName, this.value);
        if (this.invert)
            result = !result;
        return result;
    };
    SearchCriteria.parse = function (key, value) {
        var match = /^(.+?)(?::(.+))?$/.exec(key);
        if (match != null) {
            var fieldName = match[1];
            var operatorName = match[2];
            return new SearchCriteria(fieldName, operatorName, value);
        }
        return null;
    };
    ;
    return SearchCriteria;
})();
exports.SearchCriteria = SearchCriteria;
