// _js-base@0.0.1-beta@js-base/dist/SearchOperator.js
var ArgumentNullError_1 = require("./ArgumentNullError");
var SearchOperator = (function () {
    function SearchOperator(aliases, exec, parseValue, validateValue) {
        if (parseValue === void 0) { parseValue = null; }
        if (validateValue === void 0) { validateValue = null; }
        ArgumentNullError_1.ArgumentNullError.check(aliases, "aliases");
        ArgumentNullError_1.ArgumentNullError.check(exec, "exec");
        this.aliases = (Array.isArray(aliases) ? aliases : [aliases]);
        this.exec = this.wrapExec(exec);
        if (parseValue != null)
            this.parseValue = parseValue;
        else
            this.parseValue = function (value) { return value; };
        if (validateValue != null)
            this.validateValue = validateValue;
        else
            this.validateValue = function (value) { };
    }
    SearchOperator.prototype.wrapExec = function (exec) {
        return function (entity, fieldName, value) {
            if (entity.hasOwnProperty(fieldName))
                return exec(entity, fieldName, value);
            return true;
        };
    };
    SearchOperator.prototype.getName = function () {
        return this.aliases[0];
    };
    SearchOperator.Registry = {};
    return SearchOperator;
})();
exports.SearchOperator = SearchOperator;
